Issue 1281 warning on no write

mrmeku opened the pull request:
<!-- 
Thanks for contributing!
Read our contribution guidelines here: 
https://github.com/cypress-io/cypress/blob/develop/.github/CONTRIBUTING.md 
-->

<!-- Example: "Closes #1234" -->

- Closes #1281

### User facing changelog
Cypress no longer requires write access to the projectRoot, it instead will display a warning.
<!--
Explain the change(s) for every user to read in our changelog.
-->

### Additional details
- This allows for cypress inetegration with other tools such as Bazel
- It allows Cypress to be run in sandboxed or otherwise readonly fs environments

### How has the user experience changed?

#### Before

<img width="828" alt="Screen Shot 2020-05-05 at 10 35 11 AM" src="https://user-images.githubusercontent.com/1271364/81034142-2856e280-8ebc-11ea-9700-46831acfbe1d.png">

#### After

<img width="802" alt="Screen Shot 2020-05-05 at 10 25 28 AM" src="https://user-images.githubusercontent.com/1271364/81034022-bb434d00-8ebb-11ea-9d43-30063eb5a0f1.png">

### PR Tasks

<!-- 
These tasks must be completed before a PR is merged.
Delete tasks if they are not applicable. 
-->

- [x] Have tests been added/updated?
- [ ] Has the original issue been tagged with a release in ZenHub? <!-- (internal team only)-->
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->

Commit:
replace fs write error with warning

This change replaces the error with a warning
when cypress is executing tests in a readonly fs.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cypress-io/cypress?pullRequest=7126) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/cypress-io/cypress?pullRequest=7126) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: Toxicable<br/>:x: mrmeku<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cypress-io/cypress?pullRequest=7126) it.</sub>

mrmeku commented:
@jennifer-shehane Could I get a review on this? Its a continuation of a PR you were reviewing: https://github.com/cypress-io/cypress/pull/5584

jennifer-shehane commented on packages/server/lib/errors.coffee:
```diff
@@ -944,6 +944,14 @@ getMsgByType = (type, arg1 = {}, arg2, arg3) ->
 
       To avoid this error, ensure that there are no other instances of Firefox launched by Cypress running.
       """
+    when "FOLDER_NOT_ACCESSIBLE"
+      """
+      Folder #{arg1} is not writable.
+
+      Either enable write permission to this directory to enable screen shot and video collection.
+
+      Or ignore this warning if you do not need enable screen shot and video.
+      """
```
Just updating some of the warning message to make it clear on the 'why' this is necessary plus some rewording.

```suggestion
    when "FOLDER_NOT_ACCESSIBLE"
      """
      Folder #{arg1} is not writable.
      
      Writing to this directory is required by Cypress in order to store screenshots and videos.

      Enable write permissions to this directory to ensure screenshots and videos are stored. 
      
      If you don't require screenshots or videos to be stored you can safely ignore this warning.
      """
```

mrmeku commented:
@jennifer-shehane Merged your changes! Thanks for those. Is there anything else you'd like me to do before this is ready to be merged?

jennifer-shehane reviewed (changes requested):
When you add snapshots, this generates a file of the stdout output of the test you wrote. So, you need to run `SNAPSHOT_UPDATE=1 yarn test-e2e 8_read_only` to generate this file for comparison later (I've gone ahead and done this and committed it). The idea is - if something breaks, we'll see the changes in the STDOUT. 

Since the test you wrote has randomized characters - this won't work for snapshot testing. Each time you run the test, the generated output will be different. We need the test to have the same output each time the test is run.

mrmeku commented on packages/server/test/e2e/8_read_only_fs.ts:
```diff
@@ -0,0 +1,70 @@
+
+describe('e2e project dir access', function () {
+  e2e.setup()
+  const projectDir: string = path.join(os.tmpdir(), crypto.randomBytes(4).toString('hex'))
```
Ah, that's the magic incantation! Thanks so much for helping me out with that. Updating shortly

mrmeku replied:
Done!

mrmeku commented:
@jennifer-shehane Snapshots updated and ready to go

jennifer-shehane reviewed (dismissed):
There's an issue with the yarn lockfile that was committed. It's resolving to the incorrect has-binary dep. The has-binary dep should *not* have updated in your committed lockfile.

mrmeku commented:
@jennifer-shehane Sorry I didn't catch that when I was reviewing my code. It's resolved now

flotwig commented on packages/server/lib/util/settings.js:
```diff
@@ -126,7 +126,11 @@ module.exports = {
       }
 
-      // else we cannot read due to folder permissions
+      if (err.toString().startsWith('Error: EACCES: permission denied')) {
```
See how line 119 uses an object to filter the `.catch` block. You should be able to do this in a similar fashion:

```js
.catch({ code: 'EACCES' }, () => {
	return errors.warning(...)
})
```

mrmeku replied:
done! thanks for the advise

flotwig commented on packages/server/package.json:
```diff
@@ -200,4 +201,4 @@
   "optionalDependencies": {
     "registry-js": "1.8.0"
   }
-}
+}
```
revert this plz. you may want to turn on the preference to "end files with a newline" in your editor

flotwig commented on packages/server/test/e2e/8_read_only_fs.ts:
```diff
@@ -0,0 +1,55 @@
+    } catch (e) {
+      err = e
+    }
+    expect(err).to.not.be.undefined
```
```suggestion
    expect(err).to.include({ code: 'EACCES'})
```

maybe?

mrmeku replied:
Done!

flotwig commented on packages/server/test/e2e/8_read_only_fs.ts:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    // write out the project structure
+    // we cant sue one define by source code since we need to modify the file permissions on the dir and files
```
```suggestion
    // we cant use one defined by source code since we need to modify the file permissions on the dir and files
```

flotwig commented on packages/server/test/e2e/8_read_only_fs.ts:
```diff
@@ -0,0 +1,55 @@
+
+    // write out the project structure
+    // we cant sue one define by source code since we need to modify the file permissions on the dir and files
+    fs.mkdirSync(path.join(projectDir))
+
+    fs.writeFileSync(path.join(projectDir, 'cypress.json'), `{"projectId": "2pz86o"}`)
+    fs.mkdirSync(path.join(projectDir, 'cypress'))
+    fs.mkdirSync(path.join(projectDir, 'cypress', 'integration'))
+    fs.writeFileSync(path.join(projectDir, 'cypress', 'integration', 'spec.js'), `it('test',() => expect(true).to.be.true )`)
+    fs.mkdirSync(path.join(projectDir, 'cypress', 'plugins'))
+    fs.writeFileSync(path.join(projectDir, 'cypress', 'plugins', 'index.js'), `module.exports = () => {}`)
+    fs.mkdirSync(path.join(projectDir, 'cypress', 'support'))
+    fs.writeFileSync(path.join(projectDir, 'cypress', 'support', 'index.js'), `module.exports = () => {}`)
+    fs.mkdirSync(path.join(projectDir, 'cypress', 'videos'))
```
hmm, I don't think you actually need to do this project setup manually, check out my comment on `e2e.it` for a simpler way

mrmeku replied:
Done! Thanks :)

flotwig commented on packages/server/test/e2e/8_read_only_fs.ts:
```diff
@@ -0,0 +1,55 @@
+    // the dir needs executable for debugging
+    // the main test here is that it isn't writable, which would be either 6 or 7
+    // See here for more info on these numbers http://permissions-calculator.org/
+    chmodr.sync(projectDir, 0o500)
```
whoa, i didn't know this `0o` prefix was a thing, i am only familiar with using octals via the leading 0 (like `0500`)

mrmeku replied:
Yeah, you totally have to do it this way when running in strict mode
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Deprecated_octal

Learn something new about java script everyday

flotwig commented on packages/server/test/e2e/8_read_only_fs.ts:
```diff
@@ -0,0 +1,55 @@
+    expect(err).to.not.be.undefined
+  })
+
+  e2e.it('warns when unable to write to dir', {
+    project: projectDir,
+    spec: 'spec.js',
+    browser: 'chrome',
+    expectedExitCode: 0,
+    snapshot: true,
+  })
```
Test projects are automatically scaffolded in `beforeEach`, so you can do all of your `chmodr` stuff via `onRun` on one of the project dirs that already exists:

```js
  e2e.it('warns when unable to write to dir', {
    project: Fixtures.projectPath('e2e'), // copied from `test/support/fixtures/projects/e2e`
    // use a failing spec so we can see a snapshot of the screenshot failing to be written
    // in addition to the video failing to be written
    spec: 'simple_failing_spec.coffee',
    expectedExitCode: 0,
    snapshot: true,
    onRun: (exec) => {
      // ...do the `chmodr` stuff on `Fixtures.projectPath('e2e')` here
      // ...assert that EACCES is thrown here
      // now call `exec` to run the e2e test
      return exec()
      // you may need an `afterEach` call to clean up the permissions
  })
```

This way you don't need to do any of the manual project setup you have in your `beforeEach`.

mrmeku replied:
Thanks for the tip, it was driving me crazy that I couldn't do cleanup stuff before the afterEach Fixture code ran. That's why I wasn't using the built in fixture stuff

flotwig commented on packages/server/test/e2e/8_read_only_fs.ts:
```diff
@@ -0,0 +1,55 @@
+const e2e = require('../support/helpers/e2e')
+const chmodr = require('chmodr')
+
+describe('e2e project dir access', function () {
```
this is a nice, robust test. you should rename the file to `8_read_only_fs_spec.ts` to match how other e2e tests are named.

flotwig commented on packages/server/__snapshots__/8_read_only_fs.ts.js:
```diff
@@ -0,0 +1,75 @@
+  │ Spec Ran:     spec.js                                                                          │
+  └────────────────────────────────────────────────────────────────────────────────────────────────┘
+
+Warning: We failed to record the video.
+
+This error will not alter the exit code.
+
+Error: ffmpeg exited with code 1: /tmp/8_read_only_fs/cypress/videos/spec.js.mp4: Permission denied
+
+      [stack trace lines]
+
+Warning: We failed processing this video.
+
+This error will not alter the exit code.
+
+Error: ffmpeg exited with code 1: /tmp/8_read_only_fs/cypress/videos/spec.js.mp4: Permission denied
+
+      [stack trace lines]
```
would also like to see a screenshot failure here... using a spec that fails should generate a screenshot

flotwig committed:
some cleanup

- don't need to run test in a specific browser
- can combine SS + video test into one
- don't seem to need readonly_fs_spec.js

flotwig committed:
if videos folder cannot be created, just warn

flotwig commented:
Nyah... the e2e tests are failing because CI runs as root, and root doesn't get a permission denied error. Let me see how we can work around this.

flotwig committed:
try a non-root e2e test stage

flotwig reviewed (dismissed):
Nice, I did some cleanup and made it so that failure to create the `videos` directory will also not crash the test. Added a new e2e test stage so we can run the tests as non-root as well.

bahmutov commented on circle.yml:
```diff
@@ -410,168 +410,174 @@ jobs:
       - run-e2e-tests:
           browser: firefox
-          chunk: 7
+          chunk: "7"
```
why change chunking to strings?

flotwig replied:
added "non_root" chunk, and circleci has strict types for command parameters, so need to force these to be strings or else circleci complains about supplying an integer to a string parameter

bahmutov commented on packages/server/lib/errors.coffee:
```diff
@@ -944,6 +944,16 @@ getMsgByType = (type, arg1 = {}, arg2, arg3) ->
 
       To avoid this error, ensure that there are no other instances of Firefox launched by Cypress running.
       """
+    when "FOLDER_NOT_ACCESSIBLE"
```
i find it a little confusing: error is not accessible, but the real error is not writable

flotwig commented:
Note that this PR will not fix errors caused by failure to "scaffold" a new project with a "supportFile". Some work needs to be done to not scaffold supportFolder in `run` mode:

https://github.com/mrmeku/cypress/blob/b5e46dd3bc7de54b4f9234754e0228ab64a9294e/packages/server/lib/project.js#L574-L582

And with that, it will need to be made possible for individual scaffolding steps to fail and cause the default values for `config.supportFile/etc.` to change, otherwise Cypress will just end up looking for non-existent files if scaffolding fails.

bahmutov commented on packages/server/lib/util/settings.js:
```diff
@@ -121,12 +121,14 @@ module.exports = {
 
       return this._err('CONFIG_FILE_NOT_FOUND', this.configFile(options), projectRoot)
+    }).catch({ code: 'EACCES' }, () => {
+      // we cannot read due to folder permissions
```
read or write :)

bahmutov commented on packages/server/test/support/fixtures/projects/read-only-project-root/cypress/plugins/index.js:
```diff
@@ -0,0 +1,20 @@
+    err = e
+  }
+
+  expect(err).to.include({ code: 'EACCES' })
```
nice

bahmutov reviewed (dismissed):
I think this is good fix

flotwig committed:
fix comment

flotwig committed:
FOLDER_NOT_ACCESSIBLE

jennifer-shehane commented:
@flotwig @bahmutov Can this PR be approved?

jennifer-shehane commented:
@flotwig Are you saying that this PR will not fully close #1281?

flotwig commented:
@jennifer-shehane it's good to go, just waiting for those dang contributor tests to pass :roll_eyes: 

It closes #1281 since it will allow the folks in that thread to use Cypress on a read-only FS. I'll open a more specific issue about scaffolding requiring write access once this is closed.

yezhengli-Mr9 commented:
Still https://github.com/cypress-io/cypress/issues/1281
```node
See discussion and possible solutions at
npm ERR! https://github.com/cypress-io/cypress/issues/1281
npm ERR! 
npm ERR! ----------
npm ERR! 
npm ERR! Failed to access /root/.cache/Cypress:
npm ERR! 
npm ERR! EACCES: permission denied, mkdir '/root/.cache/Cypress'
```
although tried 
```
 sudo chown -R $USER:$GROUP ~/.npm
sudo chmod +x /root/.cache
```

jennifer-shehane commented:
@yezhengli-Mr9 Please open a new issue if you're encountering an issue with Cypress filling out the template.
